release-22.2: oidc: make redirect path after login configurable

dhartunian opened the pull request:
Backport 1/1 commits from #111283 

/cc @cockroachdb/release 

----

This commit adds a new cluster setting called `server.http.base_path` that's used to redirect a user logging in with OIDC SSO.

Future work will adjust other routing across the DB based on this cluster setting.

Resolves: #111242
Epic: None

Release note (ops change, ui change): A new cluster setting `server.http.base_path` is introduced that currently controls the redirection of the browser after successful login with OIDC SSO. It is expected that most customers should *not* have to adjust this. This is helpful in cases where a customer is running CRDB behind a load- balancer or proxy that serves CRDB under a subpath such as `https:// <hostname>/crdb/ `. In those cases it's necessary for the browser to redirect to `/ crdb` after login instead of `/` which has always been the hardcoded default.

----

Release justification: High-priority need for this functionality; additive only; reduced scope to OIDC only

dhartunian committed:
oidc: make redirect path after login configurable

This commit adds a new cluster setting called `server.http.base_path`
that's used to redirect a user logging in with OIDC SSO.

Future work will adjust other routing across the DB based on this
cluster setting.

Resolves: #111242
Epic: None

Release note (ops change, ui change): A new cluster setting
`server.http.base_path` is introduced that currently controls the
redirection of the browser after successful login with OIDC SSO. It
is expected that most customers should *not* have to adjust this. This
is helpful in cases where a customer is running CRDB behind a load-
balancer or proxy that serves CRDB under a subpath such as `https://
<hostname>/crdb/ `. In those cases it's necessary for the browser to
redirect to `/ crdb` after login instead of `/` which has always been
the hardcoded default.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111463)
<!-- Reviewable:end -->

dhartunian committed:
oidc: add testing of base path cluster setting

This commit adds some brute force testing of the
`server.http.base_path` cluster setting and its usage in the OIDC
configuration. It's in a separate commit because the test is a bit
branch specific and not for use in `master` as-is. In `master` we'll
be refactoring OIDC for easier testing to avoid this kind of plumbing.

However, in this branch we still want to make sure that the cluster
setting correctly results in config changes in OIDC.

Epic: None
Release note: None

knz reviewed (approved):
thank you!

dhartunian commented:
TFTR
